allennlp/modules/token_embedders/pretrained_transformer_embedder.py

```diff
@@ -1,41 +1,232 @@
+        if self._max_length is not None:
+            batch_size, num_segment_concat_wordpieces = token_ids.size()
+            token_ids, segment_concat_mask = self._fold_long_sequences(
+                token_ids, segment_concat_mask
+            )
+
+        transformer_mask = segment_concat_mask if self._max_length is not None else mask
+        # Shape: [batch_size, num_wordpieces, embedding_size],
+        # or if self._max_length is not None:
+        # [batch_size * num_segments, self._max_length, embedding_size]
+        embeddings = self.transformer_model(input_ids=token_ids, attention_mask=transformer_mask)[0]
+
+        if self._max_length is not None:
+            embeddings = self._unfold_long_sequences(
+                embeddings, segment_concat_mask, batch_size, num_segment_concat_wordpieces
+            )
+
+        return embeddings
+
+    def _fold_long_sequences(
+        self, token_ids: torch.LongTensor, segment_concat_mask: torch.LongTensor
+    ) -> Tuple[torch.LongTensor, torch.LongTensor]:
+        """
+        We fold 1D sequences (for each element in batch), returned by `PretrainedTransformerIndexer`
+        that are in reality multiple segments concatenated together, to 2D tensors, i.e.
+
+        [ [CLS] A B C [SEP] [CLS] D E [SEP] ]
+        -> [ [ [CLS] A B C [SEP] ], [ [CLS] D E [SEP] [PAD] ] ]
+        The [PAD] positions can be found in the returned `segment_concat_mask`.
+
+        # Parameters
+
+        token_ids: `torch.LongTensor`
+            Shape: [batch_size, num_segment_concat_wordpieces].
+            num_segment_concat_wordpieces is num_wordpieces plus special tokens inserted in the
+            middle, i.e. the length of: "[CLS] A B C [SEP] [CLS] D E F [SEP]" (see indexer logic).
+        segment_concat_mask: `torch.LongTensor`
+            Shape: [batch_size, num_segment_concat_wordpieces].
+
+        # Returns:
+
+        token_ids: `torch.LongTensor`
+            Shape: [batch_size * num_segments, self._max_length].
+        segment_concat_mask: `torch.LongTensor`
+            Shape: [batch_size * num_segments, self._max_length].
+        """
+        num_segment_concat_wordpieces = token_ids.size(1)
+        num_segments = math.ceil(num_segment_concat_wordpieces / self._max_length)
+        padded_length = num_segments * self._max_length
+        length_to_pad = padded_length - num_segment_concat_wordpieces
+
+        def fold(tensor):  # Shape: [batch_size, num_segment_concat_wordpieces]
+            # Shape: [batch_size, num_segments * self._max_length]
+            tensor = F.pad(tensor, [0, length_to_pad], value=0)
+            # Shape: [batch_size * num_segments, self._max_length]
+            return tensor.reshape(-1, self._max_length)
+
+        return fold(token_ids), fold(segment_concat_mask)
+
+    def _unfold_long_sequences(
+        self,
+        embeddings: torch.FloatTensor,
+        segment_concat_mask: torch.LongTensor,
```

Comment:
I'd just call this the mask.

Reply:
See above

Reply:
I want to make thinks absolutely clear by always using different terminology for different things.